src/java/main/org/apache/zookeeper/server/quorum/QuorumCnxManager.java

```diff
@@ -880,19 +876,20 @@ public void run() {
          * 
          * @return boolean  Value of variable running
          */
-        synchronized boolean finish() {
-            if(!running){
-                /*
-                 * Avoids running finish() twice. 
-                 */
-                return running;
-            }
-            running = false;            
-
-            this.interrupt();
-            threadCnt.decrementAndGet();
-            return running;
-        }
+		boolean finish() {
```

Comment:
Yes.we should keep a blance between optimization and readability.the most importance thing for this optimization is whether this function has faced a performace problem and this optimization is overdesign. Haha.self-criticism.However ,I find the shadow of this programming paradigm in zk code base [Line501~Line507](https://github.com/apache/zookeeper/blob/master/src/java/main/org/apache/zookeeper/ClientCnxn.java )